File: ui/src/types/subscription.ts
// ============================================================
// Subscription TypeScript interfaces matching Go API DTOs
// ============================================================

export type SubscriptionStatus = "active" | "paused" | "cancelled";
export type SubscriptionFrequency = "daily" | "alternate_days" | "weekly" | "monthly";

export interface SubscriptionItemResponse {
    id: number;
    variant_id: number;
    variant_name: string;
    family_name: string;
    unit: string;
    quantity: number;
}

export interface SubscriptionResponse {
    id: number;
    customer_id: number;
    customer_name: string;
    status: SubscriptionStatus;
    frequency: SubscriptionFrequency;
    start_date: string;
    end_date?: string;
    delivery_instructions?: string;
    items: SubscriptionItemResponse[];
    created_at: string;
    updated_at: string;
}

export interface SubscriptionItemRequest {
    variant_id: number;
    quantity: number;
}

export interface CreateSubscriptionRequest {
    customer_id: number;
    frequency: SubscriptionFrequency;
    start_date: string; // "YYYY-MM-DD"
    end_date?: string;
    delivery_instructions?: string;
    items: SubscriptionItemRequest[];
}

export interface UpdateSubscriptionRequest {
    frequency?: SubscriptionFrequency;
    start_date?: string;
    end_date?: string;
    delivery_instructions?: string;
    items?: SubscriptionItemRequest[];
}

export interface UpdateSubscriptionStatusRequest {
    status: SubscriptionStatus;
}

export interface SubscriptionDeliveryItemResponse {
    variant_id: number;
    variant_name: string;
    quantity: number;
    unit_price: number;
}

export interface SubscriptionDeliveryResponse {
    id: number;
    subscription_id: number;
    delivery_date: string;
    status: "delivered" | "failed" | "skipped";
    is_custom: boolean;
    items?: SubscriptionDeliveryItemResponse[];
    notes?: string;
    recorded_by?: number;
    recorded_at: string;
}

export interface RecordDeliveryRequest {
    date: string; // YYYY-MM-DD
    status: "delivered" | "failed" | "skipped";
    items?: { variant_id: number; quantity: number }[];
    notes?: string;
}

export interface DailyRosterItemResponse {
    subscription: SubscriptionResponse;
    delivery?: SubscriptionDeliveryResponse;
}

File: internal/domain/entity/subscription.go
package entity

import (
	"time"

	"github.com/shopspring/decimal"
)

// SubscriptionStatus represents the lifecycle state of a subscription
type SubscriptionStatus string

const (
	SubscriptionStatusActive    SubscriptionStatus = "active"
	SubscriptionStatusPaused    SubscriptionStatus = "paused"
	SubscriptionStatusCancelled SubscriptionStatus = "cancelled"
)

// SubscriptionFrequency represents how often deliveries occur
type SubscriptionFrequency string

const (
	SubscriptionFrequencyDaily        SubscriptionFrequency = "daily"
	SubscriptionFrequencyAlternateDays SubscriptionFrequency = "alternate_days"
	SubscriptionFrequencyWeekly       SubscriptionFrequency = "weekly"
	SubscriptionFrequencyMonthly      SubscriptionFrequency = "monthly"
)

// Subscription represents a standing recurring product order for a customer
type Subscription struct {
	ID                   int64              `json:"id"`
	CustomerID           int64              `json:"customer_id"`
	Customer             *Customer          `json:"customer,omitempty"`
	Status               SubscriptionStatus `json:"status"`
	Frequency            SubscriptionFrequency `json:"frequency"`
	StartDate            time.Time          `json:"start_date"`
	EndDate              *time.Time         `json:"end_date,omitempty"`
	DeliveryInstructions *string            `json:"delivery_instructions,omitempty"`
	Items                []SubscriptionItem `json:"items,omitempty"`
	CreatedAt            time.Time          `json:"created_at"`
	UpdatedAt            time.Time          `json:"updated_at"`
}

// SubscriptionItem represents one product line within a subscription
type SubscriptionItem struct {
	ID             int64           `json:"id"`
	SubscriptionID int64           `json:"subscription_id"`
	VariantID      int64           `json:"variant_id"`
	Variant        *ProductVariant `json:"variant,omitempty"`
	Quantity       decimal.Decimal `json:"quantity"`
}

// IsActive returns true if the subscription is currently active
func (s *Subscription) IsActive() bool {
	return s.Status == SubscriptionStatusActive
}

// IsExpired returns true if the subscription has passed its end date
func (s *Subscription) IsExpired() bool {
	if s.EndDate == nil {
		return false
	}
	return time.Now().After(*s.EndDate)
}

// DeliveryStatus represents the status of a specific daily delivery
type DeliveryStatus string

const (
	DeliveryStatusDelivered DeliveryStatus = "delivered"
	DeliveryStatusFailed    DeliveryStatus = "failed"
	DeliveryStatusSkipped   DeliveryStatus = "skipped"
)

// SubscriptionDelivery records a daily fulfillment tracking log for a subscription
type SubscriptionDelivery struct {
	ID             int64          `json:"id"`
	SubscriptionID int64          `json:"subscription_id"`
	Subscription   *Subscription  `json:"subscription,omitempty"`
	DeliveryDate   time.Time      `json:"delivery_date"`
	Status         DeliveryStatus `json:"status"`
	Notes          *string        `json:"notes,omitempty"`
	RecordedBy     *int64         `json:"recorded_by,omitempty"`
	RecordedAt     time.Time      `json:"recorded_at"`
}

// DailyRosterItem holds the aggregation of a subscription and its delivery record for a specific date
type DailyRosterItem struct {
	Subscription *Subscription         `json:"subscription"`
	Delivery     *SubscriptionDelivery `json:"delivery,omitempty"`
}
